### ISSUE TITLE
stm file and glm file

### ISSUE THREAD

**Issue by Johe-cqu** [mravanelli/pytorch-kaldi#59]
Hi,I want to use this demo to Chinese speech recognition. The example I use is thchs30 in kaldi. But there was an error in the decoding process, which was caused by the stm file not found in the score process. I compared the process of thchs30 and timit in kaldi and found that thchs30 did not generate glm file and stm file . What are the two files used for? what should I do to solve this problem?

log.log
Cannot open transcription file '/home/LLL/kaldi/egs/thchs30/s5/data/mfcc/test/stm': No such file or directory at local/timit_norm_trans.pl line 74, <M> line 61.
cp: cannot stat '/home/LLL/kaldi/egs/thchs30/s5/data/mfcc/test/glm': No such file or directory
run.pl: 10 / 10 failed, log is in /home/LLL/pytorch-kaldi-thchs30/exp/thchs30_MLP_basic_low/decode_thchs30_test_out_dnn1/scoring/log/score.*.log

score.1.log
Error: Unable to stat GLM file '/home/LLL/pytorch-kaldi-thchs30/exp/thchs30_MLP_basic_low/decode_thchs30_test_out_dnn1/scoring/glm_39phn' at /home/LLL/kaldi/tools/sctk/bin/hubscr.pl line 265.

**Comment by TParcollet** [comment]
Hi, i checked the thchs30 Kaldi recipe and saw that the decoding is based on the standard steps/decode.sh of kaldi. Therefore, it should work with our toolkit. When running the standard kaldi recipe, do you obtain the WER / PER at the end of decoding ? If so, the decoder is able to find the stm and glm files that should exist somewhere ?

**Comment by TParcollet** [comment]
Ok, so the scoring part of the thchs30 is custom. check at local/score.sh in the Kaldi recipe.

**Comment by Johe-cqu** [comment]
I am very glad that you can reply so quickly.Do I need to change the score.sh in your demo follow local/score.sh? Can you give me some tips about this?

**Comment by TParcollet** [comment]
Exactly, you have to replace this call to the right score.sh file.

**Comment by Johe-cqu** [comment]
thank you，i will try to do it.

**Comment by Subash33** [comment]
Hi, I have the same problem while decoding and scoring the test dataset. My kaldi part of the experiment only generates alignment and graphs. The data preparation I performed does not create stm and glm files. I am using TIMIT_MLP_basic as configuration file. If I create stm and glm file only for the test dataset and store in the test data directory, will the problem be solved or there are other files needed to be created?

**Comment by TParcollet** [comment]
What is your dataset ?

**Comment by Subash33** [comment]
My dataset is EMAMAE. It has both acoustic and articulatory features for
speech.

With regards,
Subash

**Comment by Johe-cqu** [comment]
In order to solve this problem, I just changed the default script of decode to the custom decode script provided in thchs30. 
But I think it won't solve your problem.

**Comment by Subash33** [comment]
> In order to solve this problem, I just changed the default script of decode to the custom decode script provided in thchs30.
> But I think it won't solve your problem.

**Comment by Subash33** [comment]
I created my own stm and glm files for my test dataset (using TIMIT as reference) then I used TIMIT_MLP_basic configuration as it is. It worked. Thanks

**Comment by Johe-cqu** [comment]
OK

**Comment by Subash33** [comment]
Hi,

I apologize in advance as the following queries maybe more kaldi related
but I am asking you here.

1. Is there any way that speaker adaptation can be carried out to the test
speakers before running the decoding and scoring on them? (My test set
speakers have different accent)
2. My transcriptions are phone sequences like TIMIT, so the WER obtained is
the PER right? Also what are these other numbers on the last line of output
showing WER? Are they insertion, deletion  substitution error counts used
to calculate WER?
3. How do one incorporate language/pronunciation models while decoding in
pytorch-kaldi framework? Or They are incorporated while computation of
alignments and graphs in kaldi?

Any help in understanding these concepts will be greatly appreciated.

With regards,
Subash

**Comment by mravanelli** [comment]
Hi, let me try to answer to you questions:

1.

*Is there any way that speaker adaptation can be carried out to the
testspeakers before running the decoding and scoring on them? (My test
setspeakers have different accent)*
One possible solution is i-vectors, x-vectors, or d-vectors. You can
compute one of these vector and concatenate with the input features.

2.


* My transcriptions are phone sequences like TIMIT, so the WER obtained
isthe PER right? Also what are these other numbers on the last line of
outputshowing WER? Are they insertion, deletion substitution error counts
usedto calculate WER?*
Yes, for phone outputs the performance reported is actually the PER. This
number is extracted from the decode*/score*. If you take a look into one of
these folder you can find in the various files more information such as the
alignment between the recognized output and the reference text, etc. I
suggest you to take a look here if you need a more detailed analysis of the
performance.

3.


*How do one incorporate language/pronunciation models while decoding
inpytorch-kaldi framework? Or They are incorporated while computation
ofalignments and graphs in kaldi?*

First of all you have to train a language model and then you have to
compile a graph that includes both language and acoustic model information.
This is done within the kaldi recipes.

Best,

Mirco

**Comment by Subash33** [comment]
Hi,

Thank you for your response. I apologize in advance as the following
queries are more generic in nature.
1. I used TIMIT_MLP_basic config file for my dataset and ended up with 62 %
PER. I know the config is not tailored for my data-set, but still do you
have any suggestions on where people generally go wrong in trying to use
these standard recipes and configurations for their own dataset? I will
consider the ivector appending tip you mentioned above. Still any
suggestions would be appreciated.
2. I am unable to get fmllr features using scripts like
"steps/nnet/make_fmllr_feats.sh" (It throws me error "Invalid feature type
[UNKNOWN]). I really need to get fmllr features to see if that would help
in reducing my PER. Is there external way of computing fmllr features if i
could not get through kaldi?
3. The decoding and scoring process for train and dev dataset, does what is
the error percentage in that, is it the WER ? If so for test set, I had to
create stm and glm file for local/score.sh to work, but i did not create
those files for train and dev set, how is the error percentage working in
those cases?


As always Thank you for your help.

With best regards,
Subash

**Comment by TParcollet** [comment]
Hi !

In general, for questions related to Kaldi, you better go into the official google-group, you will obtain more detailed and precise answers.

1. First, MLP is a bad solution to get good PER. How is the loss evolving during the training? Then, how much hours do you have ? Are the data clean or very noisy ? A lot of stuffs impacts the decoding. You should try with a bigger net to see first. IVectors, or speaker adaptation will help you to further reduce the PER, but when you're at 60% of PER, the solution is not a simple tweak, you must investigate other configurations and maybe features extraction. 

2. It will be very very hard to computer Fmllr with other toolkits and connect to Kaldi, you should try to solve your problem with generating with Kaldi.  

3. It should only be scoring test. The % reported for train and dev are within the training of the PyTorch acoustic model. Consequently, it's the loss function and not the PER. Unless you forced it ?

**Comment by Subash33** [comment]
Thank you so much.
